school/database.py

```diff
@@ -0,0 +1,514 @@
+from __future__ import annotations
+
+from sqlalchemy import (
+    BigInteger,
+    Column,
+    Integer,
+    String,
+    ForeignKey,
+    Table,
+    UniqueConstraint,
+    not_,
+)
+from sqlalchemy.orm import relationship
+from typing import Boolean, Callable, Dict, List, Optional
+
+from pie.database import database, session
+
+teachers_subjects = Table(
+    "school_dataset_teachers_subjects",
+    database.base.metadata,
+    Column(
+        "subject",
+        ForeignKey("school_dataset_subject.idx", ondelete="CASCADE"),
+        primary_key=True,
+    ),
+    Column(
+        "teacher",
+        ForeignKey("school_dataset_teacher.idx", ondelete="CASCADE"),
+        primary_key=True,
+    ),
+)
+
+
+class SubjectProgram(database.base):
+    __tablename__ = "school_dataset_subject_program"
+
+    subject_idx = Column(
+        ForeignKey("school_dataset_subject.idx", ondelete="CASCADE"),
+        primary_key=True,
+    )
+    program_idx = Column(
+        ForeignKey("school_dataset_program.idx", ondelete="CASCADE"),
+        primary_key=True,
+    )
+    year = Column(Integer, primary_key=True)
+    obligation = Column(String, primary_key=True)
+    subject = relationship("Subject", back_populates="programs")
+    program = relationship("Program", back_populates="subjects")
+
+    @staticmethod
+    def get(subject: Subject, program: Program, year: int, obligation: str):
+        query = (
+            session.query(SubjectProgram)
+            .filter_by(subject_idx=subject.idx)
+            .filter_by(program_idx=program.idx)
+            .filter_by(year=year)
+            .filter_by(obligation=obligation)
+            .one_or_none()
+        )
+
+        return query
+
+    @staticmethod
+    def add_relation(subject: Subject, program: Program, year: int, obligation: str):
+        relation = SubjectProgram(
+            subject_idx=subject.idx,
+            program_idx=program.idx,
+            year=year,
+            obligation=obligation,
+        )
+
+        session.add(relation)
+        session.commit()
+
+        return relation
+
+    def delete(self):
+        session.delete(self)
+        session.commit()
+
+
+class Teacher(database.base):
+    __tablename__ = "school_dataset_teacher"
+    __table_args__ = (
+        UniqueConstraint("school_id", "guild_id", name="school_id_guild_id_unique"),
+    )
+
+    idx = Column(Integer, primary_key=True)
+    school_id = Column(Integer)
```

Comment:
School ID ist unique for one school, but in our case the ID can be same for different guilds. This means we would have to use guild_id and school_id which results in compound primary key. That complicates a lot of things. Because of that I simplified it using generated primary key.